aten/src/ATen/SparseGCSTensorImpl.h

```diff
@@ -0,0 +1,63 @@
+#pragma once
+
+#include <ATen/Tensor.h>
+#include <ATen/TensorUtils.h>
+#include <ATen/SparseTensorUtils.h>
+#include <ATen/SparseCOOTensorImpl.h>
+#include <c10/core/TensorImpl.h>
+#include <c10/util/Exception.h>
+
+namespace at {
+  
+// TODO: since many methods in SparseCOOTensorImpl can be used by GCS sparse tensor directly
+// we probably should have some superclass between TensorImpl and GCSTensorImpl that is
+// shared between COO and GCS tensors.
+struct CAFFE2_API SparseGCSTensorImpl : public TensorImpl {
+  Tensor pointers_;
+  Tensor indices_;
+  Tensor values_;
+  Tensor reduction_;
```

Comment:
What's this? Isn't a CSR tensor normally only three tensors?

Reply:
The GCS format stores the dimensions at which you reduce the N-D tensor to a 2-D tensor in the `reduction` array. https://github.com/Quansight-Labs/rfcs/tree/pearu/rfc0005/RFC0003-sparse-roadmap#dimensionality-reduction